src/Microsoft.AspNet.OData.Shared/Query/ODataPathQueryBuilder.cs

```diff
@@ -0,0 +1,309 @@
+            }
+
+            // return queryable.Select(entity => entity.NavProp)
+            LambdaExpression selectBody =
+                Expression.Lambda(propertyExpression, param);
+            return ExpressionHelpers.Select(result, selectBody, queryable.ElementType);
+        }
+
+        /// <summary>
+        /// Transforms <paramref name="queryable"/> by selecting
+        /// the collection-valued property called <paramref name="propertyName"/>.
+        /// </summary>
+        /// <param name="queryable">The <see cref="IQueryable"/> to apply the property selection to</param>
+        /// <param name="propertyName">The name of the collection-valued property to select</param>
+        /// <returns>The transformed <see cref="IQueryable"/></returns>
+        private static IQueryable GetCollectionPropertyQuery(IQueryable queryable, string propertyName)
+        {
+            ParameterExpression param = Expression.Parameter(queryable.ElementType);
+            MemberExpression propertyExpression = Expression.Property(param, propertyName);
+
+            // for collection properties we use SelectMany instead of Select
+            // because Select would return an IQueryable<IEnumerable<TResult>>,
+            // but SelectMany flattens it to IQueryable<TResult>
+
+            // we don't check whether the property is null before advancing
+            // because that seems to cause exceptions in EF
+
+            // We expect the collection navigation property to implements IEnumerable<T>
+            // Here we extract the element type T
+            Type collectionPropertyElementType = GetEnumerableItemType(propertyExpression.Type);
+
+            // The lambda passed to queryable.SelectMany() has a signature Func<TSource, IEnumerable<TResult>>
+            // TSource corresponds to our queryable.ElementType and TResult is the element type of the nav property
+            Type delegateType = typeof(Func<,>).MakeGenericType(
+                queryable.ElementType,
+                typeof(IEnumerable<>).MakeGenericType(collectionPropertyElementType));
+            LambdaExpression selectBody =
+                Expression.Lambda(delegateType, propertyExpression, param);
+
+            // return queryable.SelectMany(entity => entity.CollectionNavProp)
+            return ExpressionHelpers
+                .SelectMany(queryable, selectBody, queryable.ElementType, collectionPropertyElementType);
+        }
+
+        /// <summary>
+        /// Gets the element type "T" of the given type
+        /// if it implements <see cref="IEnumerable{T}"/>
+        /// </summary>
+        /// <param name="enumerableType"></param>
+        /// <returns>
+        /// The element type if <paramref name="enumerableType"/> implements <see cref="IEnumerable{T}"/>
+        /// otherwise returns <paramref name="enumerableType"/> itself
+        /// </returns>
+        private static Type GetEnumerableItemType(Type enumerableType)
```

Comment:
I'd check across the repo whether there's a method that already does this...

Reply:
Alright, I'll try to find whether one exists.

Reply:
I found `TypeHelper.GetImplementedEnumerableType` which does the same thing. Thanks for the tip.